File: leetcode/0001-1000/0916-word-subsets/WordSubsets.ts
function wordSubsets(words1: string[], words2: string[]): string[] {
  const alphabets = "abcdefghijklmnopqrstuvwxyz".split("");

  const initialized = () => alphabets.reduce((accumulated, value) => {
    accumulated[value] = 0;
    return accumulated;
  }, {});

  const convert = (s: string) => s.split("").reduce(
    (accumulated, currentValue) => {
      accumulated[currentValue] += 1;
      return accumulated;
    }, initialized()
  );

  const maxConverted = (words: string[]) => words2.reduce((accumulated, value) => {
    const converted = convert(value);

    alphabets.forEach(
      c => accumulated[c] = Math.max(accumulated[c], converted[c])
    );

    return accumulated;
  }, initialized());

  const max = maxConverted(words2);

  const isValid = (w: string) => {
    const converted = convert(w);

    for (let i = 0; i < 26; i++) {
      if (max[alphabets[i]] > converted[alphabets[i]]) {
        return false;
      }
    }

    return true;
  }

  return words1.reduce(
    (accumulated, value) => {
      if (isValid(value)) {
        accumulated.push(value);
      }

      return accumulated;
    }, [] as string[]
  )
};

console.log(wordSubsets(["amazon", "apple", "facebook", "google", "leetcode"], ["e", "o"]));
console.log(wordSubsets(["amazon", "apple", "facebook", "google", "leetcode"], ["l", "e"]));

File: leetcode/0001-1000/0901-1000/0911-0920/0916-word-subsets/WordSubsets.java
import java.util.ArrayList;
import java.util.List;

public class WordSubsets {
  public static void main(String[] args) {
    String[] words1 = new String[] { "amazon", "apple", "facebook", "google", "leetcode" };
    String[] words2 = new String[] { "e", "o" };

    System.out.println(new Solution().wordSubsets(words1, words2));
  }
}

class Solution {
  int[] convert(String[] words) {
    int[] characters = new int[26];

    for (String word : words) {
      int[] temp = convert(word);

      for (int i = 0; i < 26; i++) {
        characters[i] = Math.max(characters[i], temp[i]);
      }
    }

    return characters;
  }

  int[] convert(String s) {
    int[] characters = new int[26];

    for (int i = 0; i < s.length(); i++) {
      char c = s.charAt(i);

      characters[c - 'a']++;
    }

    return characters;
  }

  boolean isValid(int[] characters, int[] max) {
    for (int i = 0; i < 26; i++) {
      if (characters[i] < max[i]) {
        return false;
      }
    }

    return true;
  }

  public List<String> wordSubsets(String[] words1, String[] words2) {
    int[] max = convert(words2);

    List<String> results = new ArrayList<>();

    for (String word : words1) {
      int[] characters = convert(word);

      if (isValid(characters, max)) {
        results.add(word);
      }
    }

    return results;
  }
}
